### ISSUE TITLE
Changes to a file belonging to a host mounted volume are not detected by nginx.

### ISSUE THREAD

**Issue by madtechsupport** [boot2docker/boot2docker#652]
I'm using boot2docker 1.3.2 on OS X 10.9.5 and after starting boot2docker and exporting the environment variables I start docker with this command:

docker run -d -p 80:80 -v "$(pwd)"/html:/usr/share/nginx/html nginx

Where in "$(pwd)/html" I have a simple index.html file. Visiting http://192.168.59.103/ displays the page for "$(pwd)/html/index.html" (as expected), but changes made to "$(pwd)/html/index.html" are not displayed when reloading the page.

To force the changes to the file to appear on the page I find it is necessary to move "$(pwd)/html/index.html" to some other filename (index.html.moved for example), reload the page for which nginx generates a "403 Fobidden" error (as no index.html exists), then move the file back to its original location (i.e. mv html/index.html.moved html/index.html), after which reloading the page will show the changes made to index.html.

Is there a better way for changes made to a file in a host mounted volume to be detected by a processes running inside the docker container?

Warren.

**Comment by haggen** [comment]
Very annoying, +1 for fixing or working around it.

**Comment by manadan999** [comment]
Its a big pain. +1

**Comment by CleanCut** [comment]
@haggen Yes, in my case getting everything to avoid using sendfile made it so I could work around the bug.  It took me weeks to find exactly what was using it, though.

**Comment by haggen** [comment]
@CleanCut Since the bug is in `vboxsf`, and it's also slow as hell, I replaced the shared folders with a nfs mount and documented the whole thing, here it is:

https://gist.github.com/haggen/f84f55e3ed2c9c930335

**Comment by felixrabe** [comment]
I just had this same issue, see nginxinc/docker-nginx#29.

**Comment by kudos** [comment]
I'm seeing the same problem with the vmwarefusion driver and docker-machine.

**Comment by gerrywastaken** [comment]
@kudos I agree, it seems that everybody (on the 20+ threads I've found for various projects) is just assuming it's a Virtualbox issue because Virtualbox hasn't responded for years. However running stat <filename> shows the file being updated the the file contents are showing as updating, so I'm wondering what it is about sendfile or nginx, that is not picking up the change.

**Comment by sleemer** [comment]
Same problem... But what I found out is:
When I update file in shared folder on my mac it reflects correctly to docker container, BUT something strange is happening with file in virtualbox vm... What I mean by strange... If, for example, I change file index.html with content 'Hello world' to 'Hi' then content of files would be:
'Hi' - /Users/username/Docker/nginx/html/index.html <--- on OS X
'Hel' - /Users/username/Docker/nginx/html/index.html <--- on VM
'Hi' - /usr/share/nginx/html/index.html <--- in the Docker container
If I change file index.html again to 'Hi again to all of you' then content of files would be:
'Hi again to all of you' - /Users/username/Docker/nginx/html/index.html <--- on OS X
'Hello world' - /Users/username/Docker/nginx/html/index.html <--- on VM
'Hi again to all of you' - /usr/share/nginx/html/index.html <--- in the Docker container

... and, of cause, nginx, with default settings, returns version of file from VM...

**Comment by jredl-va** [comment]
Just a heads up for others on this thread, I found this little tool very helpful with OSX docker development: https://github.com/brikis98/docker-osx-dev

The tool gets rsync up and running and is a great workaround for the send file issues.

**Comment by ImranAhmed** [comment]
Is this still ongoing? I hit this issue today on Windows 7, Docker Toolbox.

**Comment by yosifkit** [comment]
@ImranAhmed, yes, as far as I know it is still a bug/limitation when using VirtualBox Shared Folders.

**Comment by joe-at-startupmedia** [comment]
Still an issue for me using gulp on the host machine and hosting the index file on the client. I just run this bash script for the time being:
```bash
#!/bin/bash

### Set initial time of file
LTIME=`stat -c %Z $1`

while true
do
   ATIME=`stat -c %Z $1`

   if [[ "$ATIME" != "$LTIME" ]]
   then
     echo "" >> $1
     echo "File Saved."
     LTIME=$ATIME
   fi
   sleep 3
done
```
